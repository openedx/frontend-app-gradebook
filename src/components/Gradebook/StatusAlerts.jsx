import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { StatusAlert } from '@edx/paragon';

import { closeBanner } from '../../data/actions/grades';

export const maxCourseGradeInvalidMessage = 'Maximum course grade value must be between 0 and 100. ';
export const minCourseGradeInvalidMessage = 'Minimum course grade value must be between 0 and 100. ';

export class StatusAlerts extends React.Component {
  get isCourseGradeFilterAlertOpen() {
    const r = !this.props.isMinCourseGradeFilterValid
    || !this.props.isMaxCourseGradeFilterValid;
    return r;
  }

  get courseGradeFilterAlertDialogText() {
    let dialogText = '';
    if (!this.props.isMinCourseGradeFilterValid) {
      dialogText += minCourseGradeInvalidMessage;
    }
    if (!this.props.isMaxCourseGradeFilterValid) {
      dialogText += maxCourseGradeInvalidMessage;
    }
    return dialogText;
  }

  render() {
    return (
      <>
        <StatusAlert
          alertType="success"
          dialog="The grade has been successfully edited. You may see a slight delay before updates appear in the Gradebook."
          onClose={this.props.handleCloseSuccessBanner}
          open={this.props.showSuccessBanner}
        />
        <StatusAlert
          alertType="danger"
          dialog={this.courseGradeFilterAlertDialogText}
          dismissible={false}
          open={this.isCourseGradeFilterAlertOpen}
        />
      </>
    );
  }
}

StatusAlerts.defaultProps = {
  isMinCourseGradeFilterValid: true,
  isMaxCourseGradeFilterValid: true,
};

StatusAlerts.propTypes = {
  isMinCourseGradeFilterValid: PropTypes.bool,
  isMaxCourseGradeFilterValid: PropTypes.bool,
  // redux
  handleCloseSuccessBanner: PropTypes.func.isRequired,
  showSuccessBanner: PropTypes.bool.isRequired,
};

export const mapStateToProps = (state) => ({
  showSuccessBanner: state.grades.showSuccess,
});

export const mapDispatchToProps = {
  handleCloseSuccessBanner: closeBanner,
};

export default connect(mapStateToProps, mapDispatchToProps)(StatusAlerts);
