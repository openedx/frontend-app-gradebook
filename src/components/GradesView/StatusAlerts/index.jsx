import React from 'react';

import { Alert } from '@edx/paragon';

import useStatusAlertsData from './hooks';

export const StatusAlerts = () => {
  const {
    successBanner,
    gradeFilter,
  } = useStatusAlertsData();

  return (
    <>
      <Alert
        variant="success"
        onClose={successBanner.onClose}
        show={successBanner.show}
      >
        {successBanner.text}
      </Alert>
      <Alert
        variant="danger"
        dismissible={false}
        show={gradeFilter.show}
      >
        {gradeFilter.text}
      </Alert>
    </>
  );
};

StatusAlerts.propTypes = {};

export default StatusAlerts;
