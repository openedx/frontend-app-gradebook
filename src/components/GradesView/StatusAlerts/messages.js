import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  editSuccessAlert: {
    id: 'gradebook.GradesView.editSuccessAlert',
    defaultMessage: 'The grade has been successfully edited.  You may see a slight delay before updates appear in the Gradebook.',
    description: 'An alert text for successfully editing  a grade',
  },
  maxGradeInvalid: {
    id: 'gradebook.GradesView.maxCourseGradeInvalid',
    defaultMessage: 'Maximum course grade must be between 0 and 100',
    description: 'An alert text for selecting a maximum course grade greater than 100',
  },
  minGradeInvalid: {
    id: 'gradebook.GradesView.minCourseGradeInvalid',
    defaultMessage: 'Minimum course grade must be between 0 and 100',
    description: 'An alert text for selecting a minimum course grade less than 0',
  },
});

export default messages;
