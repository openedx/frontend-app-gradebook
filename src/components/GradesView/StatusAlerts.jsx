import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Alert } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import selectors from 'data/selectors';
import actions from 'data/actions';
import messages from './StatusAlerts.messages';

export class StatusAlerts extends React.Component {
  get isCourseGradeFilterAlertOpen() {
    return (
      !this.props.limitValidity.isMinValid
      || !this.props.limitValidity.isMaxValid
    );
  }

  get minValidityMessage() {
    return (this.props.limitValidity.isMinValid)
      ? ''
      : <FormattedMessage {...messages.minGradeInvalid} />;
  }

  get maxValidityMessage() {
    return (this.props.limitValidity.isMaxValid)
      ? ''
      : <FormattedMessage {...messages.maxGradeInvalid} />;
  }

  get courseGradeFilterAlertDialogText() {
    return (
      <>
        {this.minValidityMessage}{this.maxValidityMessage}
      </>
    );
  }

  render() {
    return (
      <>
        <Alert
          variant="success"
          onClose={this.props.handleCloseSuccessBanner}
          show={this.props.showSuccessBanner}
        >
          <FormattedMessage {...messages.editSuccessAlert} />
        </Alert>
        <Alert
          variant="danger"
          dismissible={false}
          show={this.isCourseGradeFilterAlertOpen}
        >
          {this.courseGradeFilterAlertDialogText}
        </Alert>
      </>
    );
  }
}

StatusAlerts.defaultProps = {
};

StatusAlerts.propTypes = {
  // redux
  handleCloseSuccessBanner: PropTypes.func.isRequired,
  limitValidity: PropTypes.shape({
    isMaxValid: PropTypes.bool,
    isMinValid: PropTypes.bool,
  }).isRequired,
  showSuccessBanner: PropTypes.bool.isRequired,
};

export const mapStateToProps = (state) => ({
  limitValidity: selectors.app.courseGradeFilterValidity(state),
  showSuccessBanner: selectors.grades.showSuccess(state),
});

export const mapDispatchToProps = {
  handleCloseSuccessBanner: actions.grades.banner.close,
};

export default connect(mapStateToProps, mapDispatchToProps)(StatusAlerts);
