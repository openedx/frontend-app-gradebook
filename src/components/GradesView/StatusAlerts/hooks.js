import { useIntl } from '@edx/frontend-platform/i18n';

import { actions, selectors } from 'data/redux/hooks';
import messages from './messages';

export const useStatusAlertsData = () => {
  const { formatMessage } = useIntl();

  const limitValidity = selectors.app.useCourseGradeFilterValidity();
  const showSuccessBanner = selectors.grades.useShowSuccess();
  const handleCloseSuccessBanner = actions.grades.useCloseBanner();

  const isCourseGradeFilterAlertOpen = !limitValidity.isMinValid || !limitValidity.isMaxValid;

  const validityMessages = {
    min: limitValidity.isMinValid ? '' : formatMessage(messages.minGradeInvalid),
    max: limitValidity.isMaxValid ? '' : formatMessage(messages.maxGradeInvalid),
  };

  return {
    successBanner: {
      onClose: handleCloseSuccessBanner,
      show: showSuccessBanner,
      text: formatMessage(messages.editSuccessAlert),
    },
    gradeFilter: {
      show: isCourseGradeFilterAlertOpen,
      text: `${validityMessages.min}${validityMessages.max}`,
    },
  };
};
export default useStatusAlertsData;
