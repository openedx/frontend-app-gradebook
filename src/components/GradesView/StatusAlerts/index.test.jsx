import React from 'react';
import { shallow } from 'enzyme';

import { Alert } from '@edx/paragon';

import useStatusAlertsData from './hooks';
import StatusAlerts from '.';

jest.mock('./hooks', () => jest.fn());

const hookProps = {
  successBanner: {
    onClose: jest.fn().mockName('hooks.successBanner.onClose'),
    show: 'hooks.show-success-banner',
    text: 'hooks.success-banner-text',
  },
  gradeFilter: {
    show: 'hooks.show-grade-filter',
    text: 'hooks.grade-filter-text',
  },
};
useStatusAlertsData.mockReturnValue(hookProps);

let el;
describe('StatusAlerts component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    el = shallow(<StatusAlerts />);
  });
  describe('behavior', () => {
    it('initializes component hooks', () => {
      expect(useStatusAlertsData).toHaveBeenCalled();
    });
  });
  describe('render', () => {
    test('snapshot', () => {
      expect(el).toMatchSnapshot();
    });
    test('success banner', () => {
      const alert = el.find(Alert).at(0);
      const props = alert.props();
      expect(props.onClose).toEqual(hookProps.successBanner.onClose);
      expect(props.show).toEqual(hookProps.successBanner.show);
      expect(alert.text()).toEqual(hookProps.successBanner.text);
    });
    test('grade filter banner', () => {
      const alert = el.find(Alert).at(1);
      const props = alert.props();
      expect(props.show).toEqual(hookProps.gradeFilter.show);
      expect(alert.text()).toEqual(hookProps.gradeFilter.text);
    });
  });
});
