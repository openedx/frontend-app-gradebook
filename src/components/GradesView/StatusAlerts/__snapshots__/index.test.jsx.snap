// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusAlerts component render snapshot 1`] = `
<Fragment>
  <Alert
    onClose={[MockFunction hooks.successBanner.onClose]}
    show="hooks.show-success-banner"
    variant="success"
  >
    hooks.success-banner-text
  </Alert>
  <Alert
    dismissible={false}
    show="hooks.show-grade-filter"
    variant="danger"
  >
    hooks.grade-filter-text
  </Alert>
</Fragment>
`;
